modules/tlsinfo.cpp

```diff
@@ -0,0 +1,329 @@
+static const char *TlsInfoBatchChildType = "znc.in/tlsinfo-certificate";
+
+class CTlsInfoMod : public CModule
+{
+public:
+
+#ifndef HAVE_LIBSSL
+	MODCONSTRUCTOR(CTlsInfoMod) {}
+
+	bool OnLoad(const CString&, CString& sMessage) {
+		sMessage = "Module built against install of ZNC that lacks SSL support.";
+
+		return false;
+	}
+
+#else // HAVE_LIBSSL
+
+	MODCONSTRUCTOR(CTlsInfoMod) {
+		AddHelpCommand();
+		AddCommand("Cert", static_cast<CModCommand::ModCmdFunc>(&CTlsInfoMod::PrintCertificateCommand), "[details]", "View certificate information for the active connection. Append 'details' to the 'cert' command ('cert details') to include the entire certificate chain in output.");
+		AddCommand("Cipher", static_cast<CModCommand::ModCmdFunc>(&CTlsInfoMod::PrintCipherCommand), "", "View the protocol and cipher suite used for the active connection.");
+		AddCommand("Send-Data", static_cast<CModCommand::ModCmdFunc>(&CTlsInfoMod::SendCertificateCommand), "", "Send certificate information for the active connection in an easily parsable format for application developers.");
+	}
+
+	void OnClientCapLs(CClient *mClient, SCString &mCaps) override
+	{
+		mCaps.insert(TlsInfoCap);
+	}
+
+	bool IsClientCapSupported(CClient *mClient, const CString &mCap, bool mState) override
+	{
+		return mCap.Equals(TlsInfoCap);
+	}
+
+	void PrintCipherCommand(const CString &mLine)
+	{
+		CClient *mClient;
+
+		SSL *mSSLObject;
+
+		if (GetRelevantObjects(nullptr, &mClient, nullptr, &mSSLObject) == false) {
+			return;
+		}
+
+		/* Determine the protocol used for the connection */
+		int mSSLVersion = SSL_version(mSSLObject);
+
+		CString mSSLVersionString;
+
+		if (mSSLVersion == TLS1_2_VERSION) {
+			mSSLVersionString = CString("Transport Layer Security (TLS), version 1.2");
+		} else if (mSSLVersion == TLS1_1_VERSION) {
+			mSSLVersionString = CString("Transport Layer Security (TLS), version 1.1");
+		} else if (mSSLVersion == TLS1_VERSION) {
+			mSSLVersionString = CString("Transport Layer Security (TLS), version 1.0");
+		} else if (mSSLVersion == SSL3_VERSION) {
+			mSSLVersionString = CString("Secure Sockets Layer (SSL), version 3.0");
+		} else {
+			mSSLVersionString = CString("Unknown");
+		}
+
+		/* Output cipher information */
+		CString mCipherNameString = CString(SSL_get_cipher_name(mSSLObject));
+
+		PutModule("Connection secured using " + mSSLVersionString + " with the cipher suite: " + mCipherNameString);
+	}
+
+	void PrintCertificateCommand(const CString &mLine)
+	{
+		bool mPrintCertificateParents = false;
+
+		CString sCmd = mLine.Token(1);
+
+		if (sCmd.Equals("details")) {
+			mPrintCertificateParents = true;
+		}
+
+		PresentCertificateInformation(mLine, true, mPrintCertificateParents);
+	}
+
+	void SendCertificateCommand(const CString &mLine)
+	{
+		PresentCertificateInformation(mLine, false, true);
+	}
+
+	void PresentCertificateInformation(const CString &mLine, bool mPrintCertificates = false, bool mPrintCertificateParents = false)
```

Comment:
I don't understand separation of this function and the two above.
They together do some juggling with booleans.